File: watchalert/sensitivity.py
"""Три уровня чувствительности (десктоп)."""

from __future__ import annotations

from typing import TypedDict


class SensitivityConfig(TypedDict):
    threshold: float
    min_changed_ratio: float
    label: str


LEVELS: dict[str, SensitivityConfig] = {
    "high": {
        "threshold": 3.0,
        "min_changed_ratio": 0.0,
        "label": "Сильная",
    },
    "medium": {
        "threshold": 10.0,
        "min_changed_ratio": 0.06,
        "label": "Средняя",
    },
    "low": {
        "threshold": 24.0,
        "min_changed_ratio": 0.15,
        "label": "Слабая",
    },
}


def normalize_sensitivity(value: object) -> str:
    if value in LEVELS:
        return str(value)
    if isinstance(value, (int, float)):
        n = float(value)
        if n <= 5:
            return "high"
        if n <= 14:
            return "medium"
        return "low"
    return "medium"


def get_sensitivity_config(value: object) -> SensitivityConfig:
    return LEVELS[normalize_sensitivity(value)]

File: extension/shared/sensitivity.js
/** Три уровня чувствительности (общие для расширения). */

export const SENSITIVITY_LEVELS = {
  high: {
    id: "high",
    label: "Сильная",
    hint: "на любое изменение",
    threshold: 3,
    minChangedRatio: 0,
  },
  medium: {
    id: "medium",
    label: "Средняя",
    hint: "на заметную часть",
    threshold: 10,
    minChangedRatio: 0.06,
  },
  low: {
    id: "low",
    label: "Слабая",
    hint: "только смена картинки",
    threshold: 24,
    minChangedRatio: 0.15,
  },
};

/** @param {unknown} value */
export function normalizeSensitivity(value) {
  if (value === "high" || value === "medium" || value === "low") return value;
  if (typeof value === "number" && !Number.isNaN(value)) {
    if (value <= 5) return "high";
    if (value <= 14) return "medium";
    return "low";
  }
  return "medium";
}

/** @param {unknown} level */
export function getSensitivityConfig(level) {
  return SENSITIVITY_LEVELS[normalizeSensitivity(level)];
}
